pkg/services/annotations/annotationsimpl/loki/historian_store.go

```diff
@@ -48,28 +70,229 @@ func NewLokiHistorianStore(cfg setting.UnifiedAlertingStateHistorySettings, ft f
 }
 
 func (r *LokiHistorianStore) Get(ctx context.Context, query *annotations.ItemQuery, accessResources *accesscontrol.AccessResources) ([]*annotations.ItemDTO, error) {
-	return []*annotations.ItemDTO{}, nil
+	rule := &ngmodels.AlertRule{}
+	if query.AlertID != 0 {
+		var err error
+		rule, err = getRule(ctx, r.db, query.OrgID, query.AlertID)
+		if err != nil {
+			if errors.Is(err, ErrMissingRule) {
+				return make([]*annotations.ItemDTO, 0), ErrLokiStoreInvalidQuery.Errorf("rule with ID %d does not exist", query.AlertID)
+			}
+			return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to query rule: %w", err)
+		}
+	}
+
+	logQL, err := historian.BuildLogQuery(buildHistoryQuery(query, accessResources.Dashboards, rule.UID))
+	if err != nil {
+		return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to build loki query: %w", err)
+	}
+
+	now := time.Now().UTC()
+	if query.To == 0 {
+		query.To = now.UnixMilli()
+	}
+	if query.From == 0 {
+		query.From = now.Add(-defaultQueryRange).UnixMilli()
+	}
+
+	// query.From and query.To are always in milliseconds, convert them to nanoseconds for loki
+	from := query.From * 1e6
+	to := query.To * 1e6
+
+	res, err := r.client.RangeQuery(ctx, logQL, from, to, query.Limit)
+	if err != nil {
+		return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to query loki: %w", err)
```

Comment:
in errutils there is `WithDownstream` that lets us log response error as downstream. I wonder if we can categorize this error as such.

Reply:
I don't think so, unless we consider the Loki http client as a whole downstream. `RangeQuery` can return errors before the request is made.